src/ray/core_worker/transport/direct_actor_transport.h

```diff
@@ -0,0 +1,131 @@
+      : task_id_(task_id), num_returns_(num_returns), request_(std::move(request)) {}
+
+  TaskID task_id_;
+  int num_returns_;
+  std::unique_ptr<rpc::PushTaskRequest> request_;
+};
+
+class CoreWorkerDirectActorTaskSubmitter : public CoreWorkerTaskSubmitter {
+ public:
+  CoreWorkerDirectActorTaskSubmitter(boost::asio::io_service &io_service,
+                                     gcs::RedisGcsClient &gcs_client,
+                                     CoreWorkerObjectInterface &object_interface);
+
+  /// Submit a task to an actor for execution.
+  ///
+  /// \param[in] task The task spec to submit.
+  /// \return Status.
+  Status SubmitTask(const TaskSpecification &task_spec) override;
+
+ private:
+  /// Subscribe for all actor updates.
+  Status SubscribeActorUpdates();
+
+  /// Helper function to push a task to an actor.
+  ///
+  /// \param[in] client The RPC client to send tasks to an actor.
+  /// \param[in] request The request to send.
+  /// \param[in] task_id The ID of a task.
+  /// \param[in] num_returns Number of return objects.
+  /// \return Status.
+  Status PushTask(rpc::DirectActorClient &client, const rpc::PushTaskRequest &request,
+                  const TaskID &task_id, int num_returns);
+
+  /// Treat a task as failed.
+  ///
+  /// \param[in] task_id The ID of a task.
+  /// \param[in] num_returns Number of return objects.
+  /// \param[in] error_type The type of the specific error.
+  /// \return Void.
+  void TreatTaskAsFailed(const TaskID &task_id, int num_returns,
+                         const rpc::ErrorType &error_type);
+
+  /// Handle the case that an actor becomes alive.
+  /// Note that this function doesn't take lock, the caller is expected to hold
+  /// `rpc_clients_mutex_` before calling this function.
+  ///
+  /// \param[in] actor_id Actor ID.
+  /// \param[in] ip_address The ip address of the node that the actor is running on.
+  /// \param[in] port The port that the actor is listening on.
+  /// \return Void.
+  void HandleActorAlive(const ActorID &actor_id, std::string ip_address, int port);
+
+  /// The IO event loop.
+  boost::asio::io_service &io_service_;
+
+  /// Gcs client.
+  gcs::RedisGcsClient &gcs_client_;
+
+  /// The `ClientCallManager` object that is shared by all `DirectActorClient`s.
+  rpc::ClientCallManager client_call_manager_;
+
+  /// Mutex to proect `rpc_clients_` below.
+  std::mutex rpc_clients_mutex_;
+
+  /// Map from actor ids to direct actor call rpc clients.
+  std::unordered_map<ActorID, std::unique_ptr<rpc::DirectActorClient>> rpc_clients_;
+
+  /// Map from actor ids to actor's state.
+  std::unordered_map<ActorID, gcs::ActorTableData::ActorState> actor_states_;
+
+  /// Map from actor ids to actor's location.
+  std::unordered_map<ActorID, std::pair<std::string, int>> actor_locations_;
+
+  /// Pending requests to send out on a per-actor basis.
+  std::unordered_map<ActorID, std::list<std::unique_ptr<PendingTaskRequest>>>
```

Comment:
It would be clearer to reorganize these 4 maps as one map from ActorID to an actor state class that tracks the state, client, and pending tasks.

Reply:
Thanks, yes this looks cleaner:)

Reply:
Looking at the code more closely when trying to make the change, it seems there are some differences in the 4 maps.
- for `state` and `location`, they currently track all the actors in the system, and it's natural to merge them;
- for `rpc client`, it only tracks the actors that we submit tasks to, I think after the subscribe can support individual actor, then we can have `state` and `location` to only have the actors that we submit tasks to, at that point it would be natural to merge this with them.
- for `pending tasks`, I'm feeling it's probably better to keep it separate, as we can have pending tasks before an actor is created, if we merge this together with the rest maps, we will have an entry with `state, location, rpc client` uninitialized, thus I would prefer to keep this separate.

Does this sound reasonable to you?

Reply:
Sounds good!